### PR TITLE
🔏📦 Sign distribution packages on release

### PR DESCRIPTION
**Pull request by webknjaz**
This patch integrates an invocation of the `pypi-attestations` tool[[1]] into the release automation before calling Twine. When used locally and not within a known Trusted Publishing setup, it will invoke an OAuth flow, opening a browser window, prompting the caller to authenticate via GitHub. Once authorized and browser window closed, it will proceed to signing the distribution packages[[2]] locally using Sigstore. The resulting signatures are produced in the format outlined in PEP 740 [[3]].

Twine is then called with `--attestations` which is what will actually attach the digital signatures to the release on PyPI.

[1]: https://docs.pypi.org/attestations/producing-attestations/#using-pypi-attestations
[2]: https://packaging.python.org/en/latest/glossary/#term-Distribution-Package
[3]: https://peps.python.org/pep-0740/

##### SUMMARY

$sbj.

##### ISSUE TYPE

- Maintenance Pull Request
- Packaging Pull Request

##### ADDITIONAL INFORMATION

N/A

### CODE REVIEW

**Comment by mattclay** [comment]
@webknjaz I plan on testing this with the test-ansible-release repository and test.pypi.org, although it will probably be January before I'm able to get to it.

**Comment by webknjaz** [comment]
@mattclay I tested locally up to the point of calling twine (which then fails because I didn't put any credentials in). The only thing that is untested is whether PyPI would reject the signatures (it's mentioned in the docs a bit vaguely, perhaps I can get William to spell it out more clearly…).

**Comment by webknjaz** [comment]
@woodruffw could you confirm if PyPI would accept attestations made using the “manual” way, signed by a human identity provided by GitHub's OIDC? I found that official doc a bit vague on this point.

**Comment by woodruffw** [comment]
> @woodruffw could you confirm if PyPI would accept attestations made using the “manual” way, signed by a human identity provided by GitHub's OIDC? I found that official doc a bit vague on this point.

Unfortunately, this won't (currently) work: you'll be able to produce a valid attestation with a human identity, but PyPI itself will reject the attestation at upload time. 

The reason for this is a restriction that's designed to ensure that all attestations uploaded to PyPI are also *verifiable* by PyPI: right now that means that PyPI only accepts attestations from Trusted Publisher identities. In the future however, that could be relaxed to include human identities (e.g. emails, since PyPI knows the verified emails of a project's owners/maintainers).

TL;DR: This won't currently work -- the attestation will be well-formed, but PyPI won't accept it for hosting until we figure out how best to support non-Trusted-Publishing signing identities.

**Comment by webknjaz** [comment]
@woodruffw oh, interesting. I was under the impression that PyPI *might* be accepting those already, having read the docs. My assumption was that since the identity is backed by GitHub, that might be a strong enough link. Are you implying that if there was a link between a PyPI account and a GitHub account, that might count as a verifiable identity? Is there a discussion already over in the warehouse repo? If not, we should probably start one…

@mattclay I'm marking this as a draft for the time being, since we now know that PyPI is not ready for this change.

**Comment by woodruffw** [comment]
> oh, interesting. I was under the impression that PyPI _might_ be accepting those already, having read the docs.

The docs say that PyPI only accepts Trusted Publisher identities for the moment, per [Producing attestations - Prerequisites](https://docs.pypi.org/attestations/producing-attestations/#prerequisites). However, these docs could probably be clarified, since you aren't the first person to think that other identities were already supported 🙂 -- I'm open to suggestions on how better to communicate this!

>  My assumption was that since the identity is backed by GitHub, that might be a strong enough link.

Yeah, this is an annoying case: if a user authenticates with the GitHub OIDC IdP they get an OIDC credential that's "backed" by GitHub, but doesn't actually attest to their GitHub username: it only attests that the *email* identity is one known to GitHub.

As a result, there's currently no straightforward way to do something like "I want to sign all attestations as @woodruffw" -- they can be signed as `william@yossarian.net` or similar with GitHub acting as the IdP, but that's an email identity with some interesting characteristics (since GitHub isn't actually my email provider, they're just attesting that they "know" about my email).

> Are you implying that if there was a link between a PyPI account and a GitHub account, that might count as a verifiable identity? Is there a discussion already over in the warehouse repo? If not, we should probably start one…

Yep, a link between GitHub user identities and PyPI identities is one way to solve this. Another way would be to use email identities (like above), since PyPI knows a user's verified email addresses and so does GitHub. That's currently being tracked here: https://github.com/pypi/warehouse/issues/17001

**Comment by webknjaz** [comment]
Closing since this is unlikely to ever get implemented. We can get digital signatures through #79469 instead.
